<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<hr style="height:.9px;border:none;color:#333;background-color:#333;" />

<h2>Homework - Feature Engineering</h2>
<h4>DAT-5303 | Machine Learning</h4>

<br>
<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<hr style="height:.9px;border:none;color:#333;background-color:#333;" />

<h4><b>Computation Data Analytics with Python</b></h4>

Professor Chase Kusterer

Master of Business Analytics 2024

A2: Feature Engineering (Team)

<b>Team 10: The Ogres</b> 🧌

Azharuddin Mohammed

Chih-Yu Kuo (Chewy)

Greta Tengattini

Gustavo Giopato

Martha del Río

<hr style="height:.9px;border:none;color:#333;background-color:#333;" />

<h3>Step 1: Hypothesize on Features to Engineer</h3><br>
<strong>a)</strong> Write a hypothesis on the effect you believe each new feature will have on the response variables (<em>Sale_Price</em> AND <em>log_Sale_Price</em>) in a markdown cell. (minimum 3 sentences per feature)
<br><br>
<strong>b)</strong> Give a practical, real world explanation as to why do you feel the new feature adds value. (minimum of 2 sentences per feature)

### A) FEATURES ###


**H1: total_added_space -> Supplementary convenience is expensive**

Our team assumes that the addition of supplementary, convenient architectural elements, particularly porches and attached garages, is positively related to the increase in house sales prices in the Real Estate market. Porches and garages can be extremely convenient for US homeowners for a variety of reasons, which we will explore during real-world impact analysis in section 1B.

The root of our hypothesis lies in our realization that convenient elements will not only play a functional role but could also contribute significantly to the practical utility and aesthetics levels of a property. Our team also believes that the higher the area of these environments, the higher the impact on property sale price.

We believe that convenient elements could influence buyers' positive perception of Real Estate products, which could lead to demand generation and, consequently, sales price increases from businesses. Overall, our hypothesis lies in the assumption that the existence and magnitude of convenient home features (particularly porches and garages) might indicate a positive relationship between convenience and sales price.
    
    
**H2: baths_rooms_ratio -> Full bathrooms are convenient, but they come with a cost**

Our team assumes that the higher the ratio of full bathrooms per total count of house rooms, the more expensive the house is going to be in terms of the property sales price.

An increased amount of full bathrooms in a house environment is directly linked to a higher sense of convenience and comfort since you enhance the utility of simple half bathrooms by adding a shower/bathtub. Our hypothesis lies in the assumption that this convenient, supplementary feature of a house possessing more full bathrooms per room could be related to higher property sale price.

    
**H3: 2suv_accommodation -> Storing big cars demands more structure and, consequently, more investment**

North Americans love Sports Utility Vehicles (SUVs). According to a research study conducted by Motor1 (2022), SUVs are the most preferred vehicle type by North Americans, representing 53.5% of the total utility automotive market in 2021. 

We believe that houses that can store at least 2 large SUVs (based on garage area) are more popular when it comes to clients' preferences and, consequently, demand.

A full-sized SUV is, on average, 16 feet long and has a width of approximately 7 feet (Ibiknle, 2024), which indicates that 2 units of this model would occupy a space of approximately 224 sq ft.

Based on secondary research, we identified that an average garage size is 400 sq ft (Djhamilton, 2023). We can conclude that an average-sized garage can store 2 average-sized SUVs.

Our hypothesis lies in the fact that we assume Real Estate business analysts are aware of SUV preference in the United States. Therefore, we believe that houses with garage areas that are equal to or greater than 400 sq ft and that can store at least 2 cars are related to increased sales prices in Real Estate.

    
### b. PRACTICAL EXPLANATION ###


**Feature 1: 'total_added_space'**


Real-world reasons related to the feature that have been identified as contributing to a home's appeal:
- A porch next to an attached garage creates a unified appearance that enhances curb appeal and functionality, increasing the desirability of the home (Appraisal Institute, 1994).
- Porches provide a quiet outdoor space for relaxation and entertainment, while garages provide secure storage for outdoor items, enhancing the quality of lifestyle through their combined utility (Frederick, 2005).
- Homeowners can also store outdoor gear on their porch area (4 Advantages of adding a porch to your home, 2022).
- A garage functionality goes further than simply an area where homeowners can park their cars. Individuals who own a garage can use its vast space to store goods and protect belongings. Garages could be used as a new, convenient entrance to one's house. In some specific cases, garages can also be transformed into extra rooms in case it is not needed for parking (Soda, 2023).
- A well-maintained porch can significantly increase the market price of a property, especially since outdoor spaces are increasingly valued in current living trends.


**Feature 2: 'baths_rooms_ratio'**


Real-world reasons related to the feature that have been identified as contributing to a home's appeal:
- More bathrooms means that people have more personal space and less need to wait or coordinate bathroom use, which is a significant convenience. This is particularly attractive in family homes or shared living situations where several people need to get ready at the same time (Appraisal Institute (U.S.). (1994). Appraising residential properties (2nd ed.). Appraisal Institute).
- Contemporary home buyers often seek amenities that are in keeping with a comfortable, modern lifestyle. The presence of a bathroom attached to each bedroom, often referred to as an "en-suite," is considered a desirable feature that can make a home stand out in the housing market (Gates, J. (2011, October 19). Extra bathroom key in increasing house value: Columnist Jeremy Gates rounds up all the latest property news and reveals what improvements add the most value to your home. Llanelli Star, p. 10).
- More bathrooms allow for flexible living arrangements, such as accommodating guests or renting out rooms.
- In larger homes, having bathrooms accessible from different living areas, including bedrooms, increases the functionality of the home. It eliminates the need to traverse large spaces to reach the bathroom, bedroom, and half bath.


**Feature 3: '2suv_accommodation'**


Real-world reasons related to the feature that have been identified as contributing to a home's appeal:
- North Americans like SUVs. According to Motor1 (2022), more than half of purchased vehicles in the United States are SUVs.
- An average garage size is approximately 400 sq ft (Djhamilton, 2023). This sizing is more than enough to comfortably store 2 large-sized SUV models.


In [68]:
# IMPORTING LIBRARIES
import pandas as pd                   # DATA SCIENCE ESSENTIALS
import matplotlib.pyplot as plt       # ESSENTIAL GRAPHIC OUTPUT
import seaborn as sns                 # ENHANCED GRAPHIC OUTPUT
import numpy as np                    # MATHEMATICALS ESSENTIALS
import statsmodels.formula.api as smf # REGRESSION MODELLING


# SETTING PANDAS PRINT OPTIONS
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)


# SPECIFYING FILE PATHS
file = './datasets/Ames_Housing_Dataset.xlsx'


# READING THE FILE INTO PYTHON
housing = pd.read_excel(io     = file,
                        header = 0   )

# DISPLAYING DATASET COLUMNS FOR VISUALIZATION 
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order          2930 non-null   int64  
 1   Lot_Area       2930 non-null   int64  
 2   Street         2930 non-null   object 
 3   Lot_Config     2930 non-null   object 
 4   Neighborhood   2930 non-null   object 
 5   Overall_Qual   2930 non-null   int64  
 6   Overall_Cond   2930 non-null   int64  
 7   Mas_Vnr_Area   2907 non-null   float64
 8   Total_Bsmt_SF  2929 non-null   float64
 9   First_Flr_SF   2930 non-null   int64  
 10  Second_Flr_SF  2930 non-null   int64  
 11  Gr_Liv_Area    2930 non-null   int64  
 12  Full_Bath      2930 non-null   int64  
 13  Half_Bath      2930 non-null   int64  
 14  Kitchen_AbvGr  2930 non-null   int64  
 15  TotRms_AbvGr   2930 non-null   int64  
 16  Fireplaces     2930 non-null   int64  
 17  Garage_Cars    2929 non-null   float64
 18  Garage_A

In [69]:
# SETTING 'log_Sale_Price' FOR NORMALIZATION
housing['log_Sale_Price'] = np.log1p(housing['Sale_Price'])

# REMOVING NON-NUMERIC VARIABLES FROM THE DATASET
housing = housing.drop(["Street", "Lot_Config", "Neighborhood"], axis = 1)

# CREATING A PEARSON-BASED CORRELATION MATRIX
df_corr = housing.corr().round(2)

# PRINTING PEARSON-BASED CORRELATIONS WITH SALE PRICE INFORMATION TO CHECK NORMALIZATION
# AND DATA CHANGES BASED ON OUTLIERS REMOVAL
df_corr.loc[ :, ['Sale_Price', 'log_Sale_Price'] ].sort_values(by = 'Sale_Price',
                                                                ascending = False)

Sale_Price  log_Sale_Price
Sale_Price            1.00            0.95
log_Sale_Price        0.95            1.00
Overall_Qual          0.80            0.83
Gr_Liv_Area           0.71            0.70
Garage_Cars           0.65            0.68
Garage_Area           0.64            0.65
Total_Bsmt_SF         0.63            0.63
First_Flr_SF          0.62            0.60
Full_Bath             0.55            0.58
Mas_Vnr_Area          0.51            0.45
TotRms_AbvGr          0.50            0.49
Fireplaces            0.47            0.49
Porch_Area            0.38            0.39
Half_Bath             0.29            0.31
Lot_Area              0.27            0.26
Second_Flr_SF         0.27            0.27
Pool_Area             0.07            0.05
Order                -0.03           -0.03
Overall_Cond         -0.10           -0.05
Kitchen_AbvGr        -0.12           -0.13

<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 2: Code the New Features</h3><br>
Use the code cell below to develop the features you have hypothesized.

In [70]:
## FEATURE 1 - 'total_added_space' ##

# 'total_added_space' IS A FEATURE THAT SUMS THE SQUARED FOOTAGE OF THE PORCH AND GARAGE AREAS

housing['total_added_space'] = housing['Porch_Area'] + housing['Garage_Area'] # ADDING FORMULA AND CREATING NEW FEATURE
housing.head() # CHECKING RESULTS

Order  Lot_Area  Overall_Qual  Overall_Cond  Mas_Vnr_Area  Total_Bsmt_SF  First_Flr_SF  Second_Flr_SF  Gr_Liv_Area  Full_Bath  Half_Bath  Kitchen_AbvGr  TotRms_AbvGr  Fireplaces  Garage_Cars  Garage_Area  Porch_Area  Pool_Area  Sale_Price  log_Sale_Price  total_added_space
0      1     31770             6             5         112.0         1080.0          1656              0         1656          1          0              1             7           2          2.0        528.0         272          0      215000       12.278398              800.0
1      2     11622             5             6           0.0          882.0           896              0          896          1          0              1             5           0          1.0        730.0         260          0      105000       11.561725              990.0
2      3     14267             6             6         108.0         1329.0          1329              0         1329          1          1              1             6           0          1.0        312.0         429          0      172000       12.055256              741.0
3      4     11160             7             5           0.0         2110.0          2110              0         2110          2          1              1             8           2          2.0        522.0           0          0      244000       12.404928              522.0
4      5     13830             5             5           0.0          928.0           928            701         1629          2          1              1             6           1          2.0        482.0         246          0      189900       12.154258              728.0

<br>

In [71]:
## FEATURE 2 - 'baths_rooms_ratio' ##

# 'baths_rooms_ratio' IS A FEATURE THAT CALCULATES THE RATIO OF BATHROOMS PER ROOM IN A PROPERTY

housing['baths_rooms_ratio'] = (housing['Full_Bath'] / housing['TotRms_AbvGr']).round(2)
# ADDING FORMULA, CREATING NEW FEATURE AND ROUNDING IT TO 2 DECIMAL POINTS
housing['baths_rooms_ratio'].head() # CHECKING RESULTS

0    0.142857
1    0.200000
2    0.166667
3    0.250000
4    0.333333
Name: baths_rooms_ratio, dtype: float64

<br>

In [72]:
## FEATURE 3 - '2suv_accommodation' ##

# '2suv_accommodation' IS A FEATURE WITH BOOLEAN LOGIC THAT INDICATES IF A GARAGE HAS ENOUGH SPACE TO ACCOMMODATE 2 LARGE SUVs
housing['2suv_accommodation'] = 0 # CREATING COLUMN WITH 0 AS PLACEHOLDERS

for index, row in housing.iterrows(): # CREATING DATA IMPUTATION LOGIC
    if housing.loc[index, 'Garage_Cars'] > 1 and housing.loc[index, 'Garage_Area'] > 400 : # SETTING CONDITIONS BASED ON RESEARCH
            housing.loc[index, '2suv_accommodation'] = 1 # DATA IMPUTATION LOGIC IF TRUE
    else :
        housing.loc[index, '2suv_accommodation'] = 0 # DATA IMPUTATION LOGIC IF FALSE

housing.head() # CHECKING RESULTS

Order  Lot_Area  Overall_Qual  Overall_Cond  Mas_Vnr_Area  Total_Bsmt_SF  First_Flr_SF  Second_Flr_SF  Gr_Liv_Area  Full_Bath  Half_Bath  Kitchen_AbvGr  TotRms_AbvGr  Fireplaces  Garage_Cars  Garage_Area  Porch_Area  Pool_Area  Sale_Price  log_Sale_Price  total_added_space  baths_rooms_ratio  2suv_accommodation
0      1     31770             6             5         112.0         1080.0          1656              0         1656          1          0              1             7           2          2.0        528.0         272          0      215000       12.278398              800.0           0.142857                   1
1      2     11622             5             6           0.0          882.0           896              0          896          1          0              1             5           0          1.0        730.0         260          0      105000       11.561725              990.0           0.200000                   0
2      3     14267             6             6         108.0         1329.0          1329              0         1329          1          1              1             6           0          1.0        312.0         429          0      172000       12.055256              741.0           0.166667                   0
3      4     11160             7             5           0.0         2110.0          2110              0         2110          2          1              1             8           2          2.0        522.0           0          0      244000       12.404928              522.0           0.250000                   1
4      5     13830             5             5           0.0          928.0           928            701         1629          2          1              1             6           1          2.0        482.0         246          0      189900       12.154258              728.0           0.333333                   1

<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 3: Check the Results</h3><br>
Develop a correlation matrix or a heatmap to show the linear relationships between your five new features and the response variables (<em>Sale_Price</em> AND <em>log_Sale_Price</em>).

In [73]:
# CODE HERE

features = ['total_added_space', 'baths_rooms_ratio', '2suv_accommodation', 'Sale_Price', 'log_Sale_Price'] # LISTING CREATED FEATURES AND SALE PRICE

corr_m = housing[features].corr() # DEFINING CORRELATION METHOD

corr_m.loc[['total_added_space', 'baths_rooms_ratio', '2suv_accommodation'], # RUNNING CORRELATION METHOD
           ['Sale_Price', 'log_Sale_Price']].sort_values(by = 'Sale_Price', # ASCENDING RESULTS BY 'Sale_Price'
                                                        ascending = True) 

Sale_Price  log_Sale_Price
baths_rooms_ratio     0.225738        0.266892
2suv_accommodation    0.498893        0.556307
total_added_space     0.662032        0.671264

<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 4: Explain the Results</h3><br>
Did each engineered feature perform as expected? Why or why not? (minimum of 2 sentences per engineered feature)

**Feature 1 - total_added_space**

Performance:
Our first tested hypothesis was that the existence of higher values of supplementary, convenient areas related to the porch and garage environment could be related to a higher sales price. Based on the correlation coefficients from our statistical tests (0.66 for Sale_Price and 0.67 for log_Sale_Price), we can infer that the new feature total_added_space is positively related to a higher sale price.

Final Results:
We conclude that our engineered feature worked well as our correlation metrics demonstrate a positive relationship between the newly created variable and sale price information. This could be explained by our assumption being closer to the truth - added supplementary, convenient spaces might be a direct indicator of property sales price increase. Future analyses would be needed to understand at what levels these supplementary home areas affect sales prices but correlation results go in the same direction of our primary hypothesis.


**Feature 2 - baths_room_ratio**

Performance:
Our second tested hypothesis was that the increased ratio of total full bathrooms / total rooms of a property would have a positive relationship with the Sale Price. By analyzing our correlation coefficients from our statistical tests (0.23 for Sale_Price and 0.27 for log_Sale_Price), we can infer that the new feature baths_room_ratio does not have a positive relationship with the home sale price.

Final Results:
We concluded that our engineered feature did not work well, especially when compared to our first and third engineered features. The correlation coefficient is considerably lower for Sale_Price and log_Sale_Price, which does not allow us to be confident that this feature possesses a positive relationship with the sale price. Our team was incredibly surprised to see such a low correlation value as we expected this to be one of the strongest features based on price logic and secondary research.


**Feature 3 - 2suv_accommodation**

Performance: 
Our third tested hypothesis was that the existence of homes with larger garage areas (>= 400 sq ft) that can accommodate 2 large SUVs would have a positive relationship with higher sales prices. Based on correlation coefficients from our statistical tests (0.50 for Sale_Price and 0.56 for log_Sale_Price), we can infer that the new feature 2suv_accomodation is moderately correlated to a higher sale price.


Final Results:
We concluded that our engineered feature did not work as well as expected, since we expected a much higher correlation coefficient when comparing 2suv_accommodation and Sale_Price. Although large garage areas for SUV accommodation could be linked with a higher Sale Price, our correlation results do not allow us to state with confidence that these variables have a positive relationship. Our team was surprised with these results as we expected much higher values for correlation coefficients based on secondary research and assumptions provided.

<b><h3>Reference List </b></h3>

- 4 Advantages of adding a porch to your home. (2022, February 18). https://www.baybrookremodelers.com/4-advantages-of-adding-a-porch-to-your-home

- Djhamilton. (2023, April 18). The most common 2-car garage dimensions and their uses. Hartville Outdoor Products. https://hartvilleoutdoorproducts.com/the-most-common-2-car-garage-dimensions-and-their-uses

- Frederick, G. (2005, July 22). Really, the most important room in the house: Never underestimate the value of the bathroom. Daily Townsman, p. 11.

- Gates, J. (2011, October 19). Extra bathroom key in increasing house value: Columnist Jeremy Gates rounds up all the latest property news and reveals what improvements add the most value to your home. Llanelli Star, p. 10)

- Ibiknle, D. (2024, January 29). Average car sizes: length, width, and height - Neighbor blog. Neighbor Blog. https://www.neighbor.com/storage-blog/average-car-sizes-dimensions

- Munoz, J. F. (2022, May 20). SUVs and trucks still rule US market despite higher fuel prices. Motor1.com. https://www.motor1.com/news/587230/suvs-still-rule-us-market/amp/

- Perry, N. (2023, October 12). How much value does a garage add to a house?

- Soda, T. (2023, February 8). The benefits of a garage in your new home. Royal Homes. https://www.royalhomes.com/the-benefits-of-a-garage-in-your-new-home